miniscope_io/stream_daq.py

```diff
@@ -0,0 +1,804 @@
+        eg. :class:`~miniscope_io.sdcard.BufferHeaderPositions` and :class:`~miniscope_io.sdcard.DataHeader`
+
+    """
+
+    linked_list: Tuple[int, int] = (0, 32)
+    frame_num: Tuple[int, int] = (32, 64)
+    buffer_count: Tuple[int, int] = (64, 96)
+    frame_buffer_count: Tuple[int, int] = (96, 128)
+    timestamp: Tuple[int, int] = (192, 224)
+    pixel_count: Tuple[int, int] = (224, 256)
+
+
+class MetadataHeader(BaseModel):
+    """
+    Container for FPGA header data, structured by :class:`.MetadataHeaderFormat`
+
+    """
+
+    linked_list: Any
+    """
+    Not sure what this is!
+    """
+    frame_num: int
+    buffer_count: int
+    frame_buffer_count: int
+    timestamp: int
+    pixel_count: int
+
+
+class stream_daq:
+    """
+    A combined class for reading frames from a UART and FPGA source.
+
+    .. todo::
+
+        Phil/Takuya - docstrings for stream daq: what devices these correspond to, how to configure them, usage examples, tests
+
+    """
+
+    def __init__(
+        self,
+        frame_width: int = 304,
+        frame_height: int = 304,
+        preamble: bytes = b"\x12\x34\x56",
+        header_fmt: MetadataHeaderFormat = MetadataHeaderFormat(),
+        header_len: int = 11,
+        LSB: bool = True,
+        buffer_npix: Tuple[int] = (20432, 20432, 20432, 20432, 10688),
+        pix_depth: int = 8,
+    ) -> None:
+        """
+        Constructer for the class.
+
+        Currently supports UART and FPGA source.
+
+        Parameters
+        ----------
+        frame_width : int, optional
+            Width of the frame (in pixels), by default 304.
+        frame_height : int, optional
+            Height of the frame (in pixels), by default 304.
+        preamble : bytes, optional
+            Preamble string at the beginning of every buffer header, by default b"\x12\x34\x56".
+        header_fmt : MetadataHeaderFormat, optional
+            Header format used to parse information from buffer header, by default `MetadataHeaderFormat()`.
+        header_len : int, optional
+            Length of header in (32-bit) words, by default 11.
+            This is useful when not all the variable/words in the header are defined in :class:`.MetadataHeaderFormat`.
+            The user is responsible to ensure that `header_len * 32` is larger than the largest bit position defined in :class:`.MetadataHeaderFormat` otherwise unexpected behavior might occur.
+        LSB : bool, optional
+            Whether the sourse is in "LSB" mode or not, by default True.
+            If `not LSB`, then the incoming bitstream is expected to be in Most Significant Bit first mode and data are transmitted in normal order.
+            If `LSB`, then the incoming bitstream is in the format that each 32-bit words are bit-wise reversed on its own.
+            Furthermore, the order of 32-bit words in the pixel data within the buffer is reversed (but the order of words in the header is preserved).
```

Comment:
This is, as far as I understand, what d168e696fccf7f2209d7b9649c74d8cf44f4a0ba is doing, but it really makes no sense to me. @MarcelMB @sneakers-the-rat maybe double-check this is indeed working on the microcontroller with correct firmware?

Reply:
pretty much 🤷‍♂️. I think that in future refactors we'll want to be able to coordinate with the firmware/FPGA code and resolve that bit ordering problem so that we don't need this flag, but for now this is fine

Reply:
I think I did funky bit order in firmware because:
- It's easier to handle data as 32-bit words in firmware, considering it's a 32-bit MCU.
- Some clock timing issues made receiving 32-bit asynchronous serial (UART) packets unstable, leading me to read out this data as 8-bit words on the DAQ side.

Technically, Manchester encoding should handle this problem better, and since we won't really need UART once we all get Opal Kelly boards, it might make sense to deprecate the UART function for now. This would make merging the PR easier and make it simple for me to fix the firmware in a way that makes more sense.

Reply:
sorry I guess github won't allow me to point to a single line and it's leading to some confusion. I figured the 32-bit words are reversed and that has been find for a while.

I was specifically talking about line 122 which seem to be what d168e696fccf7f2209d7b9649c74d8cf44f4a0ba is doing and it's surprising to me, since that implies on top of reversed bit order per word, the words themselves are transmitted in a chronically reversed order (and that's only true for pixels but not header??). Personally I've never seen the wireless data like this when testing with @MarcelMB so just want to make sure this is indeed the case.

Reply:
Ya ya thats this:
https://github.com/Aharoni-Lab/miniscope-io/pull/17
https://github.com/Aharoni-Lab/miniscope-io/issues/16

Reply:
Looking at that now... should that just be a full array reversal lmao. I think I was just trying to leave as much code in place as possible and not thinking about it

Reply:
it should be if, again, we can CONFIRM that bit-wise inversion of the whole buffer (minus the header) IS the intended behavior of MCU.

Reply:
fucking lmao ok this is embarassing

```python
import numpy as np

initial_array = (np.random.random((32*32,)) * 256).astype(np.uint8)

# the reverse from the current implementation
weird_reverse = np.array(
    [
        initial_array[i : i + 32][::-1]
        for i in reversed(range(0, len(initial_array), 32))
    ]
).flatten()

# how you might uh normally flip an array
regular_reverse = np.flip(initial_array)

assert np.array_equal(weird_reverse, regular_reverse)
```

Reply:
so it is just reversing the whole array. the prior implementation that just reverse the order of the 32-bit chunks while keeping the intra-chunk order was wrong, and all we did in our fix was to reverse the chunks too -- which is equivalent to just flipping the whole array but probably way slower lol

Reply:
I'm not sure why we are using the bitstring package there instead of immediately casting to a numpy array, but since the next line is just to read it into a numpy array then i say we just to the flip there lmao. let's not handle that in this PR tho, let's get this one merged and open that as its own issue

Reply:
the prior implementation worked for a few instances (including the first wireless transmission video) and is the expected behavior from what I understand.

Reply:
I think what we learned was that it appeared to work but we encountered https://github.com/Aharoni-Lab/miniscope-io/issues/16 when trying to stream an actual video (rather than just confirming that there was data being transferred) - @MarcelMB would know the history more, that's about where i came in. IDK if there was some change in the firmware, etc. but this is the kind of thing that is the reason we should have clear API contracts with tests between software/firmware/hardware components :). Go from 'this used to work' to 'we know this works because we have end-to-end tests that confirm that it works'

Reply:
anyway, that seems like something for future PRs, returning to this one - there are two items listed in the root comment that are necessary to merge this, testing on the FPGA and testing on the UART, any sense of what it would take to get that done so we can merge this and move onto the next things??